File: scala/src/main/scala/com/shekhargulati/ninetynine_problems/_03_logic_and_codes/P50.scala
package com.shekhargulati.ninetynine_problems._03_logic_and_codes

/**
  * (***) Huffman code.
  */
object P50 {

  class Tree[T](data: (Int, T), left: Tree[T], right: Tree[T])

  def huffman(fs: List[(String, Int)]): List[(String, String)] = {
    val sortedFs = fs.sortBy { case (_, f) => f }
    println(sortedFs)

    val firstTwoElements: List[(String, Int)] = sortedFs.take(2)
    val tree: Tree[String] = new Tree((firstTwoElements.map(_._2).sum, "*"), new Tree[String](firstTwoElements.head.swap, null, null), new Tree[String](firstTwoElements.last.swap, null, null))

    sortedFs.drop(2)

    ???
  }
}

File: java8/src/main/java/com/shekhargulati/ninetynine_problems/_04_binary_trees/P50.java
package com.shekhargulati.ninetynine_problems._04_binary_trees;

public class P50 {

    public static class Node<T extends Comparable<T>> implements Comparable<Node<T>> {
        T id;
        Node<T> left;
        Node<T> right;

        public Node(T id){
            this.id=id;
            this.left=null;
            this.right=null;
        }

        @Override
        public int compareTo(Node<T> object){
            return this.id.compareTo(object.id)>0?1:-1;
        }
    }


    public static boolean isTree(Node node){
        if(node==null || (node.left==null && node.left==null)){
            return true;
        }
        if(node.left!=null){
            if(node.compareTo(node.left)>0){
                if(!isTree(node.left)){
                    return false;
                }
            }else {
                return false;
            }
        }
        if(node.right!=null){
            if(node.compareTo(node.right)<0){
                if(!isTree(node.right)){
                    return false;
                }
            }else {
                return false;
            }
        }
        return true;
    }
}
